eZ/Publish/API/Repository/TrashService.php

```diff
@@ -32,15 +32,15 @@
     public function loadTrashItem($trashItemId);
 
     /**
-     * Sends $location and all its children to trash and returns the corresponding trash item.
+     * Sends $location and all its children to trash and returns the corresponding trash item, or true if the user is not allowed to read the trashed location.
      *
      * Content is left untouched.
      *
      * @throws \eZ\Publish\API\Repository\Exceptions\UnauthorizedException if the user is not allowed to trash the given location
      *
      * @param \eZ\Publish\API\Repository\Values\Content\Location $location
      *
-     * @return null|\eZ\Publish\API\Repository\Values\Content\TrashItem null if location was deleted, otherwise TrashItem
+     * @return null|true|\eZ\Publish\API\Repository\Values\Content\TrashItem null if location was deleted, otherwise TrashItem, or true if the user is not allowed to read the trashed location
```

Comment:
> TrashService is not heavily used, so could be @alongosz has some idea of how we could better model this as I assume he will dislike that it can return 3 different types now :)

It's not just different type, it's the whole thing: 
> or true if the user is not allowed to read the trashed location

Please nooo... :grinning: :rofl: 

How the `TrashItem` is used by consumer? Maybe it won't be that bad if in the case of no permissions we set Core `TrashItem::$contentInfo` to `null`?

Reply:
@alongosz Thats even worse :)

Reply:
> Thats even worse :)
Gee, thanks 😁 My idea was that trash is an action and the return value tells us how it went, one of 3 non-error outcomes: Trashed and can be read in trash, trashed and can't be read, and deleted/purged. If we should return an object, then maybe an `UnreadableTrashItem` where all properties are null (still ugly). The signal should communicate all 3 outcomes, right? We don't want to tell users that content was purged, when it was actually trashed.

Reply:
> Thats even worse :)

So if you expect `TrashItem` object, getting it instead of boolean is worse? :grinning: 
The problem indeed is that someone can rely on `TrashItem::$contentInfo` being also `ContentInfo`, so I'd say it's equally bad :)

> If we go this way we should return a different object: aka new method I guess, maybe we can continue using same signal?

+1, Seems like the cleanest way, also don't see a reason to use different signal

> If we should return an object, then maybe an UnreadableTrashItem where all properties are null (still ugly). 

Wouldn't go that direction. This would be still as bad as `true | TrashItem`.
Java style of long names would be `PermissionAwareTrashItem` which by design can have `null` properties or proper values if user has permission to read. 
But I'll understand if name would be too long for you ;)

@glye My point is to avoid methods that can return multiple types (null or type is fine since PHP 7.1). The fact that it is even possible in PHP is because PHP is broken by design. Since 7+ it goes into good direction and we should go there as well :)

Reply:
@alongosz I'm fine with this, including `PermissionAwareTrashItem`. As I understand it, it would inherit `TrashItem` so existing code works (until it tries to read contentinfo). We update what code we have, and document "BC break" which actually isn't, as André says.

Reply:
No longer relevant, using sudo instead.